packages/integrations/vercel/src/serverless/middleware.ts

```diff
@@ -0,0 +1,59 @@
+import { fileURLToPath, pathToFileURL } from 'node:url';
+import esbuild from 'esbuild';
+import { join } from 'node:path';
+import { writeFile } from '../lib/fs.js';
+
+/**
+ * It generates the Vercel Edge Middleware file.
+ *
+ * It creates a temporary file, the edge middleware, with some dynamic info.
+ *
+ * Then this file gets bundled with esbuild. The bundle phase will inline the Astro middleware code.
+ *
+ * @param astroMiddlewareEntryPoint
+ * @param outPath
+ * @returns {Promise<URL>} The path to the bundled file
+ */
+export async function generateEdgeMiddleware(
+	astroMiddlewareEntryPointPath: string,
+	outPath: string
+): Promise<URL> {
+	const code = edgeMiddlewareTemplate(astroMiddlewareEntryPointPath);
+	// Temporary file, so `esbuild` can resolve it without too much trouble.
+	// Using `stdin` option is not very optimal
+	const middlewareEdgePath = join(outPath, 'middleware-temp.mjs');
+	await writeFile(middlewareEdgePath, code);
+	// https://vercel.com/docs/concepts/functions/edge-middleware#create-edge-middleware
+	const bundledFilePath = join(outPath, 'middleware.mjs');
+	await esbuild.build({
+		target: 'es2020',
+		platform: 'browser',
+		// https://runtime-keys.proposal.wintercg.org/#edge-light
+		conditions: ['edge-light', 'worker', 'browser'],
+		external: ['astro/middleware'],
+		entryPoints: [middlewareEdgePath],
+		outfile: bundledFilePath,
+		allowOverwrite: true,
+		format: 'esm',
+		bundle: true,
+		minify: false,
+	});
+	return pathToFileURL(bundledFilePath);
+}
+
+function edgeMiddlewareTemplate(middlewarePath: string) {
+	return `
+import { onRequest } from "${middlewarePath}"
```

Comment:
Presumably if `middlewareEntryPoint` returns `URL` directly, this part could be:

```js
import { onRequest } from ${JSON.stringify(fileUrlToPath(middlewareUrl))}
```

Reply:
Oh, I noticed noticed that we do `middlewareUrl.toString()`. We use this trick somewhere else: https://github.com/withastro/astro/blob/2df9002a561a008cea0a37b52159365a3e74ae68/packages/astro/src/core/build/generate.ts#L152